gradle: fix update-deps.sh to use Nix bash

vringar opened the pull request:
## Description of changes

Fixes two issues I had with [the gradle refactor](https://github.com/NixOS/nixpkgs/pull/272380) and my external Ghidra plugins.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

I only checked that this works for my specific use case. I didn't update all Gradle packages to see if this change somehow breaks something.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

chayleaf commented on pkgs/tools/security/ghidra/build-extension.nix:
```diff
@@ -23,12 +23,13 @@ let
       ];
 
-      preBuild = ''
+      preBuild = args.preBuild or "" + ''
```
```suggestion
      preBuild = args.preBuild or "" + ''

```

chayleaf reviewed (commented):
LGTM other than a nit

chayleaf commented on pkgs/tools/security/ghidra/build-extension.nix:
```diff
@@ -23,12 +23,13 @@ let
       ];
 
-      preBuild = ''
+      preBuild = args.preBuild or "" + ''
         # Set project name, otherwise defaults to directory name
         echo -e '\nrootProject.name = "${pname}"' >> settings.gradle
       '';
```
```suggestion
      preBuild = ''
        # Set project name, otherwise defaults to directory name
        echo -e '\nrootProject.name = "${pname}"' >> settings.gradle
        ${args.preBuild or ""}
      '';
```

for example, if the `preBuild` needs to undo the `rootProject.name` changes, it's better to have user-defined `preBuild` after the default `preBuild`

vringar replied:
Nvm, this breaks my flow, as the settings.gradle file has to be created in the dir with the plugin code, but most projects don't have that at the top level.
https://github.com/vringar/ghidra-setup/blob/4e91cb063cfad510677b49ba5cf7efbfa7930ea5/extensions/lightkeeper/default.nix#L17-L19
https://github.com/vringar/ghidra-setup/blob/4e91cb063cfad510677b49ba5cf7efbfa7930ea5/extensions/ret-sync/default.nix#L19-L21

vringar commented:
Nit is addressed :) Thanks for the review

chayleaf commented:
Result of `nixpkgs-review pr 329272` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>ghidra-extensions.gnudisassembler</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>ghidra-extensions.machinelearning</li>
    <li>ghidra-extensions.sleighdevtools</li>
  </ul>
</details>

chayleaf commented on pkgs/tools/security/ghidra/build-extension.nix:
```diff
@@ -26,9 +26,11 @@ let
       '';
 
-      gradleFlags = [ "-PGHIDRA_INSTALL_DIR=${ghidra}/lib/ghidra" ];
+      gradleBuildTask = args.gradleBuildTask or "buildExtension";
```
```suggestion
```

vringar replied:
It is the default in the support file that the Ghidra docs specify as a required build file for all extension:
 https://github.com/NationalSecurityAgency/ghidra/blob/2517a3234822de74e05e642a55f79d8f3b4c5198/Ghidra/RuntimeScripts/Common/support/buildExtension.gradle#L18
 
 Unfortunately their internal extensions aren't bound by this convention

vringar replied:
I wish #315672 got merged before the refactor so we could get some more realistic users of the builder in nixpkgs.

chayleaf commented on pkgs/tools/security/ghidra/extensions/gnudisassembler/default.nix:
```diff
@@ -48,6 +48,7 @@ buildGhidraExtension {
   buildInputs = [
     zlib
   ];
+  gradleBuildTask = "build";
```
```suggestion
  gradleBuildTask = "assemble";
```

chayleaf commented:
Result of `nixpkgs-review pr 329272` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>ghidra-extensions.gnudisassembler</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>ghidra-extensions.machinelearning</li>
    <li>ghidra-extensions.sleighdevtools</li>
  </ul>
</details>

chayleaf commented on pkgs/tools/security/ghidra/extensions/gnudisassembler/default.nix:
```diff
@@ -48,6 +48,8 @@ buildGhidraExtension {
     zlib
   ];
+  gradleBuildTask = "assemble";
+  doCheck = true;
```
```suggestion
```

This will be enough to fix everything. There's no test task here, so just `assemble` is enough.

vringar commented:
`nixpkgs-review pr 329272` is passing again

chayleaf reviewed (approved):
LGTM

emilytrau commented on pkgs/development/tools/build-managers/gradle/update-deps.nix:
```diff
@@ -104,10 +105,10 @@ writeTextFile {
         --ro-bind ${toString path} ${toString path} --bind "$MITM_CACHE_CERT_DIR" "$MITM_CACHE_CERT_DIR" \
         ${builtins.concatStringsSep " " (map (x: "--setenv ${x} \"\$${x}\"") keep)} \
-        --setenv NIX_BUILD_SHELL bash ${bwrapFlags} ''${BWRAP_FLAGS:-} \
+        --setenv NIX_BUILD_SHELL ${bash}/bin/bash ${bwrapFlags} ''${BWRAP_FLAGS:-} \
```
runtimeShell can be used like so
```suggestion
        --setenv NIX_BUILD_SHELL ${runtimeShell} ${bwrapFlags} ''${BWRAP_FLAGS:-} \
```

vringar committed:
gradle: fix update-deps.sh to use Nix bash

This fixes an issue on non NixOS Linux distributions
where their system bash might be outside of
/tmp, /nix, and /run.

Thanks to @puckipedia for the suggestion

vringar committed:
ghidra: restore configurability to buildGhidraExtension

This is required as Ghidra plugins don't have a consistent
layout and have to be widely configurable

emilytrau commented:
Result of `nixpkgs-review pr 329272` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>ghidra-extensions.gnudisassembler</li>
    <li>ghidra-extensions.machinelearning</li>
    <li>ghidra-extensions.sleighdevtools</li>
  </ul>
</details>
